### PR TITLE
[LWG 19] P1673R13 A free function linear algebra interface based on the BLAS

### LINKED ISSUE #6677: [2023-11 LWG Motion 19] A free function linear algebra interface based on the BLAS
https://wiki.edg.com/pub/Wg21kona2023/StrawPolls/P1673R13.html

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #6677
Fixes cplusplus/papers#557

Issues found and fixed in separate commits:

- synopsis: missing closing `}` for namespace at end of synopsis
- synopsis: `T dot(...`: "T" was never introduced; did you mean "Scalar"?
- synopsis: `template<class Scalar, class ExecutionPolicy, ...> void symmetric_matrix_rank_k_update`: strange order of template parameters
- There is lots of mentions of "BLAS", but no expansion of the acronym, and no bibliography entry for the BLAS specification.
- [linalg.reqs.alg] p2.3 says "otherwise", but there is no preceding "if" that would match.
 - [linalg.layout.packed.obs] p3 (p13 in the paper https://wiki.edg.com/pub/Wg21kona2023/StrawPolls/P1673R13.html#wording ):  p3 would benefit from having the "let" in p4 come earlier, so that it can be used.
 - The introductory notes for [linalg.helpers.mandates] vs. [linalg.helpers.precond] are different.
 - The specification for dotc seems overly complicated. In particular, the auto-returning functions construct a type "T",
which is a bit complicated.  Instead, they could just say "return dot(conjugated(v1), v2)" and rely on the auto-returning "dot" function to do the right thing.  NOT FIXED in this pull request
 - [linalg.algs.blas1.ssq] had a duplicate definition for struct sum_of_squares_result (also present in the synopsis).
 - [linalg.algs.blas1.ssq] talks about "LAPACK", but everywhere else, we talk about "BLAS".  We need a bibliography entry for that.
 - Pervasive issue in the paper: It tries to convert "decltype(something)" to a type. You can't convert types to other types; you can only convert expressions to types. (Because the value category of the expression influences convertibility.) NOT FIXED in this pull request
 - The example scaled_matvec_2 in [linalg.algs.blas2.gemv] uses the "updating matrix-vector product" function that is defined later.  I suggest to split the example to avoid the forward reference.
 - Some functions are available with an "alpha" scalar and without. Move functions with "alpha" first.
 - [linalg.algs.blas3.rankk] p7 (for hermitian_matrix_rank_k_update) does not show "alpha" in the formula of the effects.  That feels like a bug.
 - [linalg.algs.blas3.rank2k] p6 (for hermitian_matrix_rank_2k_update) uses ^T, not ^H, in the formula.  Is that correct?
 - There are a few cross-references to [linalg.algs], which doesn't exist. I've switched them to [linalg].

### CODE REVIEW

**Comment by Dani-Hub** [comment]
@jensmaurer The definition of _imag-if-needed_ for sub-bullet (1.2) is incorrect, I assume that

otherwise, `((void)E, T)`.

should be either

otherwise, `((void)E, T{})`.

as the paper says or 

otherwise, `((void)E, T())`.

**Comment by jensmaurer** [comment]
@Dani-Hub, thanks for the hint:

> The definition of imag-if-needed for sub-bullet (1.2) is incorrect

Fixed; missing backslashes in LaTeX.

**Comment by burblebee** [comment]
@JohelEGP @jwakely (or any other volunteers) This one is huge - how about we split this review up between us? This is looks to be about fair:
* 28-28.9.2
* 28.9.3-28.9.13
* 28.9.14-end

I'll take the last one, 28.9.14-end.

**Review by burblebee — CHANGES_REQUESTED** [review]
not done, but probably done for the night - here's what I have so far...

File: source/numerics.tex [hunk 1]
```diff
@@ -10397,3 +10398,5060 @@
...
+\effects
+Computes $y = A x$.
+\end{itemdescr}
+
+\begin{example}
+\begin{codeblock}
+constexpr size_t num_rows = 5;
+constexpr size_t num_cols = 6;
+
+// y = 3.0 * A * x
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13661]: Hm... `operator*` is not provided for the involved types, so formatting this as C++ code throughout isn't correct.
Formatting y, A, x as C++ code would work, but then this is almost indistinguishable from all-code.

@tkoeppe , what do you think?

File: source/numerics.tex [hunk 2]
```diff
@@ -10397,3 +10398,5060 @@
...
+  matrix_vector_product(scaled(3.0, A), x, y);
+}
+
+// y = 3.0 * A * x + 2.0 * y
+void scaled_matvec_2(mdspan<double, extents<size_t, num_rows, num_cols>> A,
+  mdspan<double, extents<size_t, num_cols>> x, mdspan<double, extents<size_t, num_rows>> y) {
+  matrix_vector_product(scaled(3.0, A), x, scaled(2.0, y), y);
+}
+
+// z = 7.0 times the transpose of A, times y
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13667]: see above

File: source/numerics.tex [hunk 3]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{example}
+\begin{codeblock}
+constexpr size_t num_rows = 5;
+constexpr size_t num_cols = 6;
+
+// y = 3.0 * A * x
+void scaled_matvec_1(mdspan<double, extents<size_t, num_rows, num_cols>> A,
+  mdspan<double, extents<size_t, num_cols>> x, mdspan<double, extents<size_t, num_rows>> y) {
+  matrix_vector_product(scaled(3.0, A), x, y);
+}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13665]: I find all of this rather verbose. Maybe we should invent type aliases for these.
`using Matrix = mdspan<double, extents<size_t, num_rows, num_cols>>;`

@tkoeppe , what do you think?

File: source/numerics.tex [hunk 4]
```diff
@@ -10397,3 +10398,5060 @@
...
+                               InMat A, InVec1 x, InVec2 y, OutVec z);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+These functions performs an updating matrix-vector product.
+
+\pnum
+\effects
+Computes $z = y + A x$.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13690]: No, this is exactly the way it should be. We describe a mathematical operation (matrix-vector multiply and vector-vector-add).  There is no corresponding C++ code you could write easily (absent the facilities defined in the paper).

[linalg.general] p1.4 makes the connection between the math variables and the function parameter names.

File: source/numerics.tex [hunk 5]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{itemdescr}
+
+\pnum
+These functions perform an overwriting symmetric matrix-vector product,
+taking into account the \tcode{Triangle} parameter
+that applies to the symmetric matrix \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes $y = A x$.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13768]: see above

File: source/numerics.tex [hunk 6]
```diff
@@ -10397,3 +10398,5060 @@
...
+
+\begin{itemdescr}
+\pnum
+These functions perform an updating symmetric matrix-vector product,
+taking into account the \tcode{Triangle} parameter
+that applies to the symmetric matrix \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes $z = y + A x$.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13788]: see above

File: source/numerics.tex [hunk 7]
```diff
@@ -10397,3 +10398,5060 @@
...
+
+\pnum
+\expects
+\begin{itemize}
+\item
+\tcode{A.extent(0)} equals \tcode{A.extent(1)},
+\item
+\tcode{\exposid{multipliable}(A,x,y)} is \tcode{true}, and
+\item
+\tcode{\exposid{addable}(x,y,z)} is \tcode{true} for those overloads that take a \tcode{z} parameter.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13831]: fixed throughout

File: source/numerics.tex [hunk 8]
```diff
@@ -10397,3 +10398,5060 @@
...
+
+\begin{itemdescr}
+\pnum
+These functions perform an overwriting Hermitian matrix-vector product,
+taking into account the \tcode{Triangle} parameter
+that applies to the Hermitian matrix \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes $y = A x$.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13855]: see above --- correct and intended as-is.

File: source/numerics.tex [hunk 9]
```diff
@@ -10397,3 +10398,5060 @@
...
+\complexity
+\bigoh{\tcode{y.extent(0)} \times \tcode{A.extent(1)}}
+\end{itemdescr}
+
+\begin{itemdecl}
+template<@\exposconcept{in-matrix}@ InMat, class Triangle, class DiagonalStorage,
+         @\exposconcept{in-vector}@ InVec1, @\exposconcept{in-vector}@ InVec2, @\exposconcept{out-vector}@ OutVec>
+  void triangular_matrix_vector_product(InMat A, Triangle t, DiagonalStorage d,
+                                        InVec1 x, InVec2 y, OutVec z);
+template<class ExecutionPolicy, @\exposconcept{in-matrix}@ InMat, class Triangle, class DiagonalStorage,
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13990]: I've done so in the synopsis; in the \itemdecls, it's more compact this way, and we're allowed to reflow there.

File: source/numerics.tex [hunk 10]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{itemdecl}
+template<@\exposconcept{in-vector}@ InVec1, @\exposconcept{in-vector}@ InVec2, @\exposconcept{inout-matrix}@ InOutMat>
+  void matrix_rank_1_update(InVec1 x, InVec2 y, InOutMat A);
+template<class ExecutionPolicy, @\exposconcept{in-vector}@ InVec1, @\exposconcept{in-vector}@ InVec2, @\exposconcept{inout-matrix}@ InOutMat>
+  void matrix_rank_1_update(ExecutionPolicy&& exec, InVec1 x, InVec2 y, InOutMat A);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+These functions perform a nonsymmetric nonconjugated rank-1 update.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:14197]: It seems to be just some normal (math) prose utterance.

File: source/numerics.tex [hunk 11]
```diff
@@ -10397,3 +10398,5060 @@
...
+otherwise, equivalent to:
+\begin{codeblock}
+matrix_rank_1_update(std::forward<ExecutionPolicy>(exec), x, conjugated(y), A);
+\end{codeblock}
+\end{itemize}
+\end{itemdescr}
+
+\rSec3[linalg.algs.blas2.symherrank1]{Symmetric or Hermitian Rank-1 (outer product) update of a matrix}
+
+\begin{note}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:14258]: fixed

File: source/numerics.tex [hunk 12]
```diff
@@ -10397,3 +10398,5060 @@
...
+These functions correspond to the BLAS functions
+\tcode{xSYR}, \tcode{xSPR}, \tcode{xHER}, and \tcode{xHPR}.
+They have overloads taking a scaling factor \tcode{alpha},
+because it would be impossible to express the update
+$A = A - x x^T$ otherwise.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:14263]: see [linalg.general] p1.4 and general math notation conventions, which omits multiplication symbols by default.

File: source/numerics.tex [hunk 13]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{itemize}
+\item
+\tcode{A.extent(0)} equals \tcode{A.extent(1)}, and
+\item
+\tcode{A.extent(0)} equals \tcode{x.extent(0)}.
+\end{itemize}
+
+\pnum
+\complexity
+\bigoh{\tcode{x.extent(0)} \times \tcode{x.extent(0)}}
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:14295]: fixed

File: source/numerics.tex [hunk 14]
```diff
@@ -10397,3 +10398,5060 @@
...
+\pnum
+These functions perform
+a Hermitian rank-1 update of the Hermitian matrix \tcode{A},
+taking into account the \tcode{Triangle} parameter
+that applies to \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes $A'$ such that
+$A' = A + \alpha x x^H$, where the scalar $\alpha$ is \tcode{alpha},
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:14360]: no, see above

File: source/numerics.tex [hunk 15]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{itemdescr}
+\pnum
+These functions perform
+a symmetric rank-1 update of the symmetric matrix \tcode{A},
+taking into account the \tcode{Triangle} parameter
+that applies to \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes a matrix $A'$ such that $A' = A + x x^T$
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex]: no, see above

File: source/numerics.tex [hunk 16]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{itemdescr}
+\pnum
+These functions perform
+a symmetric rank-2 update of the symmetric matrix \tcode{A},
+taking into account the \tcode{Triangle} parameter
+that applies to \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes $A'$ such that $A' = A + x y^T + y x^T$ and
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:14442]: no, see above

**Comment by JohelEGP** [comment]
> @JohelEGP @jwakely (or any other volunteers) This one is huge - how about we split this review up between us? This is looks to be about fair:
> 
>     * 28-28.9.2
> 
>     * 28.9.3-28.9.13
> 
>     * 28.9.14-end
> 
> 
> I'll take the last one, 28.9.14-end.

In that case, I'll take the first one.

File: source/numerics.tex [hunk 17]
```diff
@@ -10397,3 +10398,5060 @@
...
+then the elements of \tcode{X} are valid but unspecified.
+
+\pnum
+\complexity
+$O($ \tcode{B.extent(0)} $\cdot$ \tcode{B.extent(1)} $\cdot$ \tcode{A.extent(1)} $)$
+\begin{note}
+Since the triangular matrix is on the right,
+the desired \tcode{divide} implementation
+in the case of noncommutative multiplication
+is mathematically equivalent to $x y^{-1}$,
```
Comment (burblebee) [inline source/numerics.tex:15266]: you have a space here in the Latex source (between x and y), yet the rendering looks good (i.e. no space in pdf, so multiplication assumed) - this is better than having a space in the pdf (the space makes it look like 2 separate expressions).

... there still needs to be an association between /x/ and \tcode{x} tho

**Review by burblebee — CHANGES_REQUESTED** [review]
Still have issues with:
* \tcode vs. italics used to represent the same variables (in paper also),
* spaces between variables in math expressions (looks like separate expressions, but without space as in paper, the string looks like a single variable), and
* missing "is \tcode{true}" after some expressions (in paper also)

File: source/numerics.tex (hunk 8 above)
[the remark this answers is not in the document]
Comment (burblebee) [reply source/numerics.tex:13875]: fyi - resolving similar comments for all but a few issues pointed out to reduce noise (please apply resolution to above issue to all cases however)

File: source/numerics.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex:13708]: In math, variables usually have a single letter, possibly with some decoration (e.g. superscript or subscript indices etc.) That makes it unambiguous to omit the multiplication sign.

In contrast, C++ has multi-letter variables, and thus we need a multiplication sign, but that is the C++ multiplication, not the math multiplication, but the former doesn't exist for the types in question.

File: source/numerics.tex (hunk 17 above)
Comment (jensmaurer) [reply source/numerics.tex:15266]: Is this comment trying to express more than your general discomfort with implicitly equating math variables with C++ function parameters via [linalg.general] p1?  If so, I don't get it; please try to rephrase.

Side note: LaTeX math mode has special rendering rules, and spaces in the source don't necessarily turn into spaces in the resulting formula.

File: source/numerics.tex [hunk 18]
```diff
@@ -10397,3 +10398,5060 @@
...
+\begin{itemize}
+\item
+If \tcode{InMat} has \tcode{layout_blas_packed} layout,
+then the layout's \tcode{Triangle} template argument has
+the same type as the function's \tcode{Triangle} template argument;
+\item
+\tcode{\exposid{possibly-multipliable}<InOutMat, InMat, InOutMat>()}
+is \tcode{true}; and
+\item
+\tcode{\exposid{compatible-static-extents}<InMat, InMat>(0,1)}.
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/numerics.tex]: Fixed throughout.

**Comment by jensmaurer** [comment]
@jwakely , is there any specific text in the paper (that I'm currently overlooking) that would associate input math variables with same-named C++ function parameters? [linalg.general] p1.4 talks about output variables only.

File: source/numerics.tex (hunk 17 above)
Comment (burblebee) [reply source/numerics.tex:15266]: Yes, but maybe moot now? The spaces in the Latex source were sometimes rendered in the pdf, making some mathy expressions look like two separate expressions (e.g. $A' = A + x x^T$ became "A' = A + x x^T" with a space between the two x's). But I can't find where those cases are in the latest pdf, so you fixed them already. I looked through all the "Effects: Computes" and they look fine now.

Comment (jensmaurer) [reply source/numerics.tex:15266]: Interesting, I'm not aware of any changes in that area, or even of the possibility of having an influence (absent `\ ` or similarly gross things).

File: source/numerics.tex [hunk 19]
```diff
@@ -10397,3 +10398,5060 @@
...
+a triangular solve,
+taking into account the \tcode{Triangle} and \tcode{DiagonalStorage} parameters
+that apply to the triangular matrix \tcode{A}\iref{linalg.general}.
+
+\pnum
+\effects
+Computes a vector $x'$ such that $b = A x'$,
+and assigns each element of $x'$ to the corresponding element of $x$.
+If no such $x'$ exists,
+then the elements of \tcode{x} are valid but unspecified.
```
Comment (burblebee) [inline source/numerics.tex:14103]: This x should not be \tcode for consistency

**Review by burblebee — CHANGES_REQUESTED** [review]
Just 1 last required change

**Comment by jwakely** [comment]
> @jwakely , is there any specific text in the paper (that I'm currently overlooking) that would associate input math variables with same-named C++ function parameters? [linalg.general] p1.4 talks about output variables only.

Erm, possibly not. We have similar problems in [rand], see #4295 for an old attempt to fix it there.

File: source/numerics.tex (hunk 1 above)
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/numerics.tex:13748]: We should probably write the code as code, but the whole expression as in maths. We can either use `\times` or `\cdot` or just juxtaposition for the various kinds of multiplication.

But we can clean that up later, too.

File: source/numerics.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/numerics.tex:13752]: Yes, that would be helpful. Let's do that as a follow-up.

**Comment by tkoeppe** [comment]
Thank you all very much for the reviews, and special thanks to Jens for the drafting! I'll merge this now, and we can perform follow-up cleanup afterwards. Thanks again!

**Pull request merged**